hat/app/org/hatdex/hat/api/controllers/RichData.scala

```diff
@@ -70,342 +103,622 @@ class RichData @Inject() (
+        val disabled = for {
+          _ <- dataDebitService.dataDebitDisable(dataDebitId)
+          debit <- dataDebitService.dataDebit(dataDebitId)
+        } yield debit
+
+        disabled
+          .andThen(
+            dataEventDispatcher
+              .dispatchEventMaybeDataDebit(DataDebitOperations.Disable()))
+          .map {
+            case Some(debit) => Ok(Json.toJson(debit))
+            case None        => BadRequest(Json.toJson(Errors.dataDebitDoesNotExist))
+          }
     }
 
-  private object Errors {
-    def dataDebitDoesNotExist = ErrorMessage("Not Found", "Data Debit with this ID does not exist")
-    def dataDebitNotFound(id: String) = ErrorMessage("Not Found", s"Data Debit $id not found")
-    def dataDebitNotEnabled(id: String) = ErrorMessage("Bad Request", s"Data Debit $id not enabled")
-    def dataDebitMalformed(err: Throwable) = ErrorMessage("Bad Request", s"Data Debit request malformed: ${err.getMessage}")
-    def dataDebitBundleMalformed(id: String, err: Throwable) = ErrorMessage("Data Debit Bundle malformed", s"Data Debit $id active bundle malformed: ${err.getMessage}")
+  private def makeData(
+    namespace: String,
+    endpoint: String,
+    orderBy: Option[String],
+    ordering: Option[String],
+    skip: Option[Int],
+    take: Option[Int])(implicit db: HATPostgresProfile.api.Database): Future[Result] = {
+    val dataEndpoint = s"$namespace/$endpoint"
+    val query =
+      Seq(EndpointQuery(dataEndpoint, None, None, None))
+    val data = dataService.propertyData(
+      query,
+      orderBy,
+      ordering.contains("descending"),
+      skip.getOrElse(0),
+      take.orElse(Some(defaultRecordLimit)))
+    data.map(d => Ok(Json.toJson(d)))
+  }
 
-    def bundleNotFound(bundleId: String) = ErrorMessage("Bundle Not Found", s"Bundle $bundleId not found")
+  def getContractData(
+    namespace: String,
+    endpoint: String,
+    orderBy: Option[String],
+    ordering: Option[String],
+    skip: Option[Int],
+    take: Option[Int]): Action[AnyContent] =
+    UserAwareAction.async { implicit request => //    SecuredAction(
+      {
+
+        val ret = for {
+          isOk <- isValidContractRequest(request.body.asJson)
+          ret <- if (isOk)
+            makeData(namespace, endpoint, orderBy, ordering, skip, take)
+          else Future.successful(BadRequest)
+        } yield ret
+
+        ret
+      }
+    }
 
-    def dataUpdateMissing(message: String) = ErrorMessage("Data Missing", s"Could not update records: $message")
-    def dataDeleteMissing(message: String) = ErrorMessage("Data Missing", s"Could not delete records: $message")
-    def dataLinkMissing(message: String) = ErrorMessage("Data Missing", s"Could not link records: $message")
+  private def bodyOptToContractDataRequest(
+    bodyOpt: Option[JsValue]): Option[ContractDataRequest] = {
+    for {
+      bodyAsJson <- bodyOpt
+      token <- refineV[NonEmpty]((bodyAsJson \ "token").as[String]).toOption
```

Comment:
I would define a case class for the expected JSON body and then leverage play-json to handle marshalling. For example, see `HatJsonFormats` trait in the HAT Client library and how it is being used. Not sure how it will interplay with Refined though.

Reply:
Comment with link reads much better 👍

Reply:
I have this refactored to sort this out, just not pushed yet.